NSS: memcache_timeout=0

mzidek-rh opened the pull request:
This is PR is continuation of:
https://github.com/SSSD/sssd/pull/416

about new semantics of memcache_timeout=0

lslebodn commented:
Please keep change minimal. As it was agreen on team meeting

mzidek-rh commented:
I really do believe that it is better to move the memcache related code from nss_process_init to a separate function and I think it is better to do moves of this kind before downstream rebases.

There are several reason as to why I moved the code there:
- there is already too much memcache related code in the nss_process_init and moving it to a static function makes things more readable
- I do not like 'cutting' the nss_process_init function with one check for timeout==0 and then returning from the function, because there is some code after the memcache initialization that should be executed and also we could shoot ourselves in the future by adding non memcache related code after this return that would not be executed when the timeout is zero. So the only way to solve this is by putting the whole memcache related stuff into one big if (not good) or move it to a function. I know that there is also the possibility that you used in your PR to put the condition in the mmap_cache_init function, because IMO it makes more sense to put this to a function where we read the timeout value (or all memcache init values to be precise) because it is obvious what effects these options have to memcache from reading just this part of the code. This benefit will IMO even more obvious once we add the options to change memcache sizes.

All in all I do not think it is beneficial for us to go with the minimum lines of code this time.

mzidek-rh commented:
I removed the changes requested label without changing the code (explanation is in my previous comment).

jhrozek commented:
I think there might also be confusion around what the "minimal" approach as agreed on the meeting is. I don't think it was meant as trying to go for the minimal amount of lines changed, but at least I understood it as us dropping PR #390 and introducing this one. Even with the split of a block of code into a new function, the patch is not very risky.

In addition, we agreed to not introduce a configure time option if I recall correctly (and the meeting notes explicitly say "without conditionals")

jhrozek commented:
The patch seems to work OK, but there is no test..I wrote a quick one in https://github.com/jhrozek/sssd/tree/review if you think it would be enough, can you squash it into your patch and resubmit?

jhrozek commented on src/man/sssd.conf.5.xml:
```diff
@@ -974,12 +974,19 @@ fallback_homedir = /home/%u
                             Default: 300
                         </para>
                         <para>
+                            WARNING: Disabling in-memory cache will
```
'the in-memory cache'.

jhrozek commented on src/responder/nss/nsssrv.c:
```diff
@@ -277,6 +277,12 @@ static int setup_memcaches(struct nss_ctx *nctx)
     }
 
+    if (memcache_timeout == 0) {
+        DEBUG(SSSDBG_TRACE_FUNC,
```
Shouldn't this be CONF_SETTINGS?

jhrozek commented:
Please see the note about the test and the nitpick in the man page and the debug level..

lslebodn commented:
Minimal means minimal. And yesterday there were just two ways which were discussed
https://github.com/SSSD/sssd/pull/390/files
and
https://github.com/SSSD/sssd/pull/390#issuecomment-334448236

mzidek-rh commented:
@jhrozek I included the test you wrote. It LGTM, I am currently running the intgcheck suite locally to see if it passes. The only change I made to the test was that I added a blank line to suppress pep8 warning.

The other nitpicks are also addressed.

mzidek-rh commented:
Btw. feel free to add your Signed-of-by tag to the that I merged with your test.

lslebodn commented:
Thank you very much for changes.

To person who will push the patches. Please do not mention my name anywhere in commit messages. I do not want to have anything related with this misfeature.

mzidek-rh commented:
@lslebodn about the signed-off-by, I was talking to @jhrozek

lslebodn commented:
On (20/10/17 08:31), mzidek-rh wrote:
>@lslebodn about the signed-off-by, I was talking to @jhrozek 
>

We also use Review-by, and I can even see Tested-by
in git log

LS

mzidek-rh commented:
@lslebodn well I know that, I just thought you are reacting to that signed-off-by comment I made. I see it probably was not the case.

fidencio commented:
Ack! I'm firing a CI build and will add the "Changes Requested" label as soon as I get the results.

Whoever decides to push this series, please, also add @pbrezina, @sumit-bose, @jhrozek as reviewers.

fidencio commented:
Well, as there's no internal CI and CentOS Ci passed, I'm adding the "Accepted" label to this PR.

jhrozek commented:
* master:
1becbb7bec29a3d418d8f19fc52433cf86bcf395                                                                                                                                                                                                    
ffe29e570a9e885c2f0061c34bb6be2bbd6ab9e4
878b0d42aca5839fdc1d97a68ce181e280f1ed7b

lslebodn commented:
5 days ago there were just two commit:
 c3d9de2
 31dab23

And 3 patches were pushed. Including one with refactoring. May I know why?

It is not minimal chage.

mzidek-rh commented:
@lslebodn not sure what you mean. There were always three patches in my branch. One that moves code, one that adds functionality and one with man page changes.

lslebodn commented:
I did so many compromises and requested minimal change. https://github.com/SSSD/sssd/pull/420#issuecomment-338213789. Without any refactoring.

It seems that github is far from optional tool for review.  https://github.com/SSSD/sssd/pull/414#issuecomment-338160535

Anyway you might expect that some downstream distributions will not accept this solutions.

mzidek-rh commented:
Ok, I think I see where the confusion comes from. After your comment "Minimal means minimal" I force pushed the changes that Jakub requested and GH showed in the comment section only the patches that were changed, not the one with the moved code. After that, your next comment looked like you are not 100% happy, but you do accept the we plan to go with the patches. I see it was because you though the first patch was removed. Well, that is unfortunate misunderstanding, but I still think the change is not big and the patch that moves the code to function makes it better. I gave arguments for that in 3rd comment in this thread.

I think it is better to not rely on the review functionality here on GH (the same goes for pagure) and always look for the content in the actual branches, which may not be that convenient, but is more reliable.

As for downstream distributions not accepting these patches, the only reason I could see is problematic backporting, but I think that at this point it is not a problem to backport these patches to any currently supported downstream distro. On the contrary, it could cause unnecessary conflicts in the future if the downstream distribution decided to not include these patches, because then this part of the code would differ more as we go forward.
